Step3 - 계산기록

wefu0000 opened the pull request:
안녕하세요
3단계 계산기록 구현했습니다.
이전 피드백 강의에서 공유된 내용을 따라하다보니
의도하지 않게 4단계도 일부 적용해 버렸네요;;
많은 조언 부탁드립니다.

wefu00 committed:
feat: 수식 저장 기능 구현

1. Data 모듈 추가, Room을 이용한 수식 저장/가져오기 기능 추가
2. 저장한 수식 보여주는 기능 구현

wefu00 committed:
feat: repository 구조 변경, DB, Entity 네임 변경

wefu00 committed:
test: db test (room, robolectric)

wefu00 committed:
build: ktlint 체크 제외 옵션 추가

wefu00 committed:
build: liveData lint 오류 수정

wefu00 committed:
build: lint 오류 수정

AnKyungMoo commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorActivity.kt:
```diff
@@ -21,13 +37,31 @@ class CalculatorActivity : AppCompatActivity() {
         }
+
+        viewModel.memoryItems.observe(this) { items ->
+            with(binding.recyclerView.adapter as MemoryAdapter) {
```
`binding.recyclerView.adapter`가 현재는 `MemoryAdapter`밖에 없는데요..!
다른 어댑터가 같이 관리될 수 있다고하면 해당 코드는 크래시가 발생하는 코드가 됩니다.
조금 더 안정성을 챙길 수 있는 방법은 무엇이 있을까요?

wefu0000 replied:
adapter를 멤버 변수로 선언해서 사용하게 수정했습니다. 
여러 좋은 방법이 더 있을꺼 같아, 의견 부탁드려요~

AnKyungMoo replied:
멤버 변수로 초기화하여 사용하는 것도 좋은 방법입니다 👍 

그 외의 방법으로는 2가지 정도 생각이 나는데요..!
1. (binding.recyclerView.adapter as? MemoryAdapter) 
2. if (binding.recyclerView.adapter is MemoryAdapter) 를 한 뒤, as이용 

1번의 경우 adapter가 MemoryAdapter로 캐스트 되지 않는다면 null로 반환되어 그 뒤에 코드를 실행시키지 않을 수 있습니다.

2번의 경우 binding객체가 var이기때문에 언제든지 변경이 될 수 있습니다. 따라서 스마트캐스트가 지원되지 않아 추가적으로 as를 이용해야하는데요. 결국 스마트캐스트가 지원되지 않는 이유와 같이 멀티스레드 환경같은 상황에서 binding이 변경되는 경우 as도 크래시가 발생할 수 있어 안정성을 챙길 수는 없습니다.

제가 코드를 직접 작성했다면 상인님이 하신 것 처럼 멤버 변수로 어댑터를 선언하거나 1번 as?를 사용할 것 같네요ㅎㅎ

2번은 참고차 같이 작성하였습니다.

wefu0000 replied:
1번 예제의  as? 에 대한 동작에 대해서 알게 되었습니다!!!
좋은 의견 감사합니다.

AnKyungMoo commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/MemoryAdapter.kt:
```diff
@@ -0,0 +1,33 @@
+import camp.nextstep.edu.calculator.domain.data.Memory
+
+class MemoryAdapter(
+    var items: List<Memory> = emptyList()
```
```suggestion
    val items: List<Memory> = emptyList()
```

var로 받지 않는 것이 더 혼동을 줄일 수 있을 것 같네요..!

wefu0000 replied:
코드를 수정하려고 하다보니 
계산식이 추가될때마다 items가 변경해야되어야 합니다.
val로 바꾸게 되면 계산식이 추가될때마다 Adapter를 따로 생성해야하는데
val로 바꿀수 있는 좋은 방법이 있을까요? ㅠㅠ

AnKyungMoo replied:
현재는 MemoryAdapter에 있는 값을 다른 객체들에서 변경할 수 있는 상태입니다.
이런 경우 캡슐화가 잘 진행되어있지 않게 되네요..!
객체가 가지고있는 속성들은 객체 스스로만 변경할 수 있고, 다른 객체들은 접근만 할 수 있도록 설계하는 것이 객체지향적인 관점에 더 잘 어울립니다.

힌트로는..! backing properties에 관련해서 찾아보시면 좋을 것 같습니다 :)

wefu0000 replied:
items를 업데이트 하기 위해 updateItems 함수를 만들고, 
items는 private으로 변경하였습니다.
리뷰어님 의도에 맞게 수정한건지 궁금합니다.

AnKyungMoo commented on calculator-data/src/androidTest/java/camp/nextstep/edu/calculator/data/MemoryEntityDbTest.kt:
```diff
@@ -0,0 +1,47 @@
+
+        memoryDao.insert(memoryEntity)
+        val resultList = memoryDao.getAll()
+        val actual = resultList[resultList.size - 1]
```
Collections.last() 를 이용해 볼 수 있습니다.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/last.html

wefu0000 replied:
좋습니다!
수정하였습니다

AnKyungMoo commented on calculator-data/src/main/java/camp/nextstep/edu/calculator/data/MemoryDatabase.kt:
```diff
@@ -0,0 +1,29 @@
+
+    companion object {
+        private const val databaseName = "memory.db"
+        private var Instance: MemoryDatabase? = null
```
```suggestion
        private var instance: MemoryDatabase? = null
```

코틀린에서 변수명의 첫글자는 소문자로 사용하는 것이 더욱 컨벤션에 맞습니다.

wefu0000 replied:
네 맞습니다!
수정했습니다.

AnKyungMoo commented on calculator-data/src/main/java/camp/nextstep/edu/calculator/data/MemoryEntity.kt:
```diff
@@ -0,0 +1,24 @@
+
+@Entity
+data class MemoryEntity(
+    @PrimaryKey val id: UUID = UUID.randomUUID(),
```
Nit: autoGenerate를 이용하면 randomUUID를 사용하지 않아도 괜찮습니다.

wefu0000 replied:
네 좋습니다.
수정했습니다.

AnKyungMoo commented on calculator-domain/src/main/java/camp/nextstep/edu/calculator/domain/data/Memory.kt:
```diff
@@ -0,0 +1,6 @@
+package camp.nextstep.edu.calculator.domain.data
+
+data class Memory(
```
이름을 조금 더 명확한 것으로 변경하면 좋을 것 같아요.
현재는 클래스명만 확인해서는 어떤 의도인지 알기 어렵습니다.

wefu0000 replied:
네 좋습니다.
ResultExpression 으로 수정하였습니다.

AnKyungMoo commented on calculator-data/src/main/java/camp/nextstep/edu/calculator/data/repository/MemoryLocalRepository.kt:
```diff
@@ -0,0 +1,31 @@
+) : MemoryRepository {
+    private val memoryDao: MemoryDao? = memoryDatabase?.resultDao()
+
+    override suspend fun getMemoryList(): List<Memory> {
```
Q: Dao에서는 suspend함수로 만들지 않았는데, Repository에서만 suspend로 설정해준 이유가 있을까요?
현재 정상적으로 동작하는지 궁금합니다.

wefu0000 replied:
Room 을 통해서 DB 접근을하기 위해서는 main 스레드에서 접근이 불가능해서 
비동기 처리를 강제하기 위해 suspend 를 설정했습니다.
suspend 함수를 만들기보다는 db에 접근하는 내부 repository에서 비동기 처리를 해줘야할까요?

AnKyungMoo replied:
아닙니다..! 제가 이 부분은 살짝 착각했네요 ㅎ;
dao에서 suspend가 되어있지 않아도 정상 실행이 되는지 잠시 혼동이 있었어요..ㅠ
이 코멘트는 넘어가셔도 좋습니다..!

AnKyungMoo commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorActivity.kt:
```diff
@@ -4,13 +4,29 @@ import android.os.Bundle
     private lateinit var binding: ActivityCalculatorBinding
 
-    private val viewModel: CalculatorViewModel by viewModels()
+    private val viewModel: CalculatorViewModel by viewModels {
+        object : ViewModelProvider.Factory {
+            override fun <T : ViewModel> create(modelClass: Class<T>): T {
+                return if (modelClass.isAssignableFrom(CalculatorViewModel::class.java)) {
+                    CalculatorViewModel(
+                        memoryRepository = DataInjector.provideMemoryRepository(this@CalculatorActivity.application)
+                    ) as T
+                } else {
+                    throw IllegalArgumentException()
+                }
+            }
+        }
+    }
```
Nit: 물론 한 번만 사용할 코드지만 충분히 확장될 수 있는 코드기때문에 따로 클래스를 선언해두어도 좋을 것 같아요..!
꼭 반영하지는 않으셔도 좋습니다 :)

wefu0000 replied:
좋습니다. 
ViewModel 내부에 클래스를 따로 선언하게 수정했습니다.

AnKyungMoo reviewed (changes requested):
전체적으로 많이 잘 반영해주셨네요 👍 
다음 step 넘어가기 전에 코멘트 몇가지 남겨두었습니다.
수정 후 다시 요청 주세요~!

wefu00 committed:
fix: 리뷰 내용 반영

UUID.randomUUID() -> autoGenerate

wefu00 committed:
fix: 리뷰 내용 반영

코틀린에서 컨벤션에 맞게 변수명 변경

wefu00 committed:
fix: 리뷰 내용 반영

collections last 사용

wefu00 committed:
fix: 리뷰 내용 반영

명확한 네이밍으로 변경: Memory -> ResulExpression

wefu00 committed:
fix: 리뷰 내용 반영

recyclerview의 adapter를 형변환하지않고 선언하여 사용

wefu00 committed:
fix: 리뷰 내용 반영

ViewModelProvider.Factory ViewModel 클래스에 선언

wefu00 committed:
fix: ktlincheck 수정

wefu0000 commented:
안녕하세요. 
리뷰 내용 반영했습니다.
리뷰 해주신 내용중에 반영하기 어려운게 있어 댓글 달았습니다.
많은 조언 부탁드립니다

AnKyungMoo commented on calculator-data/src/main/java/camp/nextstep/edu/calculator/data/MemoryEntity.kt:
```diff
@@ -0,0 +1,24 @@
+    var expression: String = "",
+
+    @ColumnInfo(name = "result")
+    var result: String = ""
```
MemoryEntity의 값은 한 번 설정되고나면 추가적으로 변경되지 않습니다.
이런 경우 var를 사용하기보단 val을 사용하는 것이 좋아요..!

추가적으로 `MemoryEntity`도 `ResultExpression`처럼 이름이 명확하게 변경되면 좋을 것 같습니다!

wefu0000 replied:
네 변경하였습니다
좋은 의견 감사합니다!

AnKyungMoo commented on calculator-data/src/test/java/camp/nextstep/edu/calculator/data/ResultExpressionRepositoryTest.kt:
```diff
@@ -0,0 +1,42 @@
+
+            resultExpressionRepository.addResultExpression(resultExpression)
+            val memoryList = resultExpressionRepository.getResultExpressionList()
+            val actual = memoryList[memoryList.size - 1]
```
여기 부분도 last를 사용할 수 있습니다!

wefu0000 replied:
네 변경하였습니다
좋은 의견 감사합니다!

AnKyungMoo commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -36,14 +63,40 @@ class CalculatorViewModel : ViewModel() {
+                throw IllegalArgumentException()
+            }
+        }
+    }
```
CalculatorViewModel 안에 클래스를 만드신 이유가 있을까요?!

wefu0000 replied:
CalculatorViewModel 생성할때 필요한 Factory 클래스라서
CalculatorViewModel 내부에 class로 선언했는데요
별도 클래스로 만드는게 나을까요?

AnKyungMoo replied:
ViewModelFactory는 Factory패턴을 적용하여 ViewModel 객체를 만들어주는 역할을 하고있습니다.
현재는 CalculatorViewModel 한가지 밖에 없지만, 다른 뷰모델이 타입이 생겼을 때 같은 팩토리에서 케이스에 따라 다른 뷰모델을 생성해주어야할 경우가 있습니다.

특별한 이유가 있지 않다면, 확장성의 개념때문에 별도의 클래스로 만드는 것이 더욱 좋습니다!

AnKyungMoo commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorActivity.kt:
```diff
@@ -4,13 +4,18 @@ import android.os.Bundle
     private lateinit var binding: ActivityCalculatorBinding
 
-    private val viewModel: CalculatorViewModel by viewModels()
+    private val resultExpAdapter: ResultExpAdapter by lazy { ResultExpAdapter() }
```
lazy로 해당 객체를 초기화하신 이유가 있을까요?!

추가적으로 expression이라는 단어가 짧지는 않지만, 줄여서 사용하기보단 풀로 풀어서 사용하는 것을 권장드립니다.
축약어의 사용은 약속되지 않은 사람들간에 이해를 저하시키는 요인이라고 생각해요..!

wefu0000 replied:
lazy로 선언하면 실제 해당 객체가 필요할때 생성되는 걸로 알고 있어 사용하였습니다.
Adapter 생성이 해당 화면에서 꼭 필요하긴해서 바로 대입하여 생성해도 되는데 어떠한 방법이 더 좋을까요?
Adapter 네임은 expression로 변경하였습니다.
좋은 의견 감사합니다.

AnKyungMoo replied:
현재 상황에서는 둘 다 크게 차이가 없습니다 ㅎ;
다만 Adapter에 추가로 초기화해야할 점이나, 설정해주어야할 것이 없기 때문에 타이밍에 맞춰서 선언하는 lazy를 사용하지 않아도 괜찮습니다.

저는 객체가 생성되는 타이밍에 현재 가지고 있는 속성값들을 기준으로 설정되어야할 점이 있는 경우가 아니라면 lazy를 이용하진 않습니다.

AnKyungMoo reviewed (changes requested):
질문 주신 내용 확인하여 답변 남겼습니다..!
수정해주신 내용 포함하여 추가적으로 코멘트 더 남겼는데요~
다음 step은 금방 마칠 수 있으니..! 이번 step에서 조금만 더 힘을 내보면 좋을 것 같습니다 :)

관련하여 다른 질문이 생긴다면 추가 코멘트 남겨주시거나 DM 보내주세요!!

wefu00 committed:
fix: 리뷰 내용 반영

1. entity 명 ResultExpression 변경, var -> val 사용
2. db test 에서 배열 마지막 접근 -> last() 사용
3. ResultExpAdapter -> ResultExpressionAdapter 변경
4. adapter에서 var items 접근을 updateItem을 만들어서 외부에서 설정하게 하고 private 으로 접근 불가능하게 변경

wefu0000 commented:
안녕하세요.
리뷰 내용 반영했습니다.
항상 좋은 리뷰 감사합니다.
리뷰 해주신 내용중에 잘 반영한게 맞는지 궁금한 내용이 있어 댓글 달았습니다.
많은 조언 부탁드립니다

wefu00 committed:
fix: 불필요한 파라미터 제거

AnKyungMoo commented on calculator-app/src/main/res/layout/item_result.xml:
```diff
@@ -1,27 +1,36 @@
+<layout>
+    <data>
+        <variable
+            name="memory"
```
ResultExpression로 변수명이 같이 바뀌면 좋겠네요 😅

AnKyungMoo reviewed (approved):
고생하셨습니다~!!
질문주셨던 부분에 대해서는 추가 코멘트 남겨두었습니다~ 같이 확인해주세요!

다음 step으로 진행하기 위해 머지하도록 할게요!
